pkg/routeagent_driver/handlers/calico/ippool_handler.go

```diff
@@ -0,0 +1,198 @@
+)
+
+type calicoIPPoolHandler struct {
+	event.HandlerBase
+	restConfig      *rest.Config
+	client          *calicocs.Clientset
+	remoteEndpoints map[string]*submV1.Endpoint
+	isGateway       bool
+}
+
+var logger = log.Logger{Logger: logf.Log.WithName("CalicoIPPool")}
+
+func NewCalicoIPPoolHandler(restConfig *rest.Config) event.Handler {
+	return &calicoIPPoolHandler{
+		restConfig:      restConfig,
+		remoteEndpoints: map[string]*submV1.Endpoint{},
+	}
+}
+
+func (h *calicoIPPoolHandler) GetNetworkPlugins() []string {
+	return []string{cni.Calico}
+}
+
+func (h *calicoIPPoolHandler) GetName() string {
+	return "Calico IPPool handler"
+}
+
+func (h *calicoIPPoolHandler) Init() error {
+	var err error
+
+	h.client, err = calicocs.NewForConfig(h.restConfig)
+
+	return errors.Wrap(err, "error initializing Calico clientset")
+}
+
+func (h *calicoIPPoolHandler) RemoteEndpointCreated(endpoint *submV1.Endpoint) error {
+	h.remoteEndpoints[endpoint.Name] = endpoint
+	if !h.isGateway {
+		logger.V(log.TRACE).Info("Ignore RemoteEndpointCreated event (node isn't Gateway)")
+		return nil
+	}
+
+	err := h.createIPPool(endpoint)
+
+	return errors.Wrap(err, "failed to handle RemoteEndpointCreated event")
+}
+
+func (h *calicoIPPoolHandler) RemoteEndpointRemoved(endpoint *submV1.Endpoint) error {
+	delete(h.remoteEndpoints, endpoint.Name)
+
+	if !h.isGateway {
+		logger.V(log.TRACE).Info("Ignore RemoteEndpointRemoved event (node isn't Gateway)")
+		return nil
+	}
+
+	err := h.deleteIPPool(endpoint)
+
+	return errors.Wrap(err, "failed to handle RemoteEndpointRemoved event")
+}
+
+func (h *calicoIPPoolHandler) TransitionToNonGateway() error {
+	logger.Info("TransitionToNonGateway")
+
+	h.isGateway = false
+
+	return nil
+}
+
+func (h *calicoIPPoolHandler) TransitionToGateway() error {
+	logger.Info("TransitionToGateway")
+
+	retErr := error(nil)
+	h.isGateway = true
+
+	for _, endpoint := range h.remoteEndpoints {
+		err := h.createIPPool(endpoint)
+		if err != nil {
+			logger.Warningf("Failed to create ippool %s", endpoint.GetName())
+
+			retErr = err
+		}
+	}
+
+	return retErr
+}
+
+func (h *calicoIPPoolHandler) Stop(uninstall bool) error {
+	if !uninstall || !h.isGateway {
+		return nil
+	}
+
+	logger.Info("Uninstalling Calico IPPools used for Submariner")
+
+	labelSelector := labels.SelectorFromSet(map[string]string{submarinerIPPool: "true"}).String()
+	err := h.client.ProjectcalicoV3().IPPools().DeleteCollection(context.TODO(), metav1.DeleteOptions{},
+		metav1.ListOptions{LabelSelector: labelSelector})
+
+	if err != nil && !apierrors.IsNotFound(err) {
+		return errors.Wrapf(err, "Failed to delete Calico IPPools using labelSelector %q", labelSelector)
+	}
+
+	logger.Infof("Successfully delete Calico IPPools using labelSelector %q", labelSelector)
+
+	return nil
+}
+
+func (h *calicoIPPoolHandler) createIPPool(endpoint *submV1.Endpoint) error {
+	subnets := cidr.ExtractIPv4Subnets(endpoint.Spec.Subnets)
+	retErr := error(nil)
```

Comment:
I think for completeness we should return an [aggregate error](https://github.com/kubernetes/apimachinery/blob/c9b3b3a37189129d0d80b894dc5d393ade9d3b79/pkg/util/errors/errors.go#L46) instead of just returning the last one, ie:

```
import errorutils "k8s.io/apimachinery/pkg/util/errors"
...
var retErrors []error

...
retErrors = append(retErrors, ...)
...

return errorutils.NewAggregate(retErrors)
```